### PR TITLE
implement deal status v1

### PR DESCRIPTION
**Pull request by dirkmc**

### CODE REVIEW

File: db/deals.go [hunk 1]
```diff
@@ -76,6 +73,8 @@ func newDealAccessor(db *sql.DB, deal *types.ProviderDealState) *dealAccessor {
 			"Length":                &fieldDef{f: &deal.Length},
 			"Checkpoint":            &ckptFieldDef{f: &deal.Checkpoint},
 			"Error":                 &fieldDef{f: &deal.Err},
+			// Needed so the deal can be looked up by signed proposal cid
+			"SignedProposalCID": &signedPropFieldDef{prop: deal.ClientDealProposal},
```
Comment (aarshkshah1992) [inline db/deals.go:77]: Please can we add a test for this field to the db/tests ?

File: storagemarket/lp2pimpl/net.go [hunk 2]
```diff
@@ -89,26 +95,30 @@ func NewDealClient(h host.Host, options ...DealClientOption) *DealClient {
...
+		prov:     prov,
+		fullNode: fullNodeApi,
 	}
 	return p
 }
 
 func (p *DealProvider) Start() {
 	p.host.SetStreamHandler(DealProtocolID, p.handleNewDealStream)
 	p.host.SetStreamHandler(mktssm.AskProtocolID, p.handleNewAskStream)
+	p.host.SetStreamHandler(mktssm.DealStatusProtocolID, p.handleNewDealStatusStream)
```
Comment (aarshkshah1992) [inline storagemarket/lp2pimpl/net.go:115]: Can you create an issue to remove this handler from `go-fil-markets` then after the integration ?

File: storagemarket/lp2pimpl/net.go [hunk 3]
```diff
@@ -167,7 +177,120 @@ func (p *DealProvider) handleNewAskStream(s network.Stream) {
...
-		log.Errorf("failed to write ask response: %s", err)
+		log.Warnw("failed to write ask response", "err", err)
 		return
 	}
 }
+
+func (p *DealProvider) handleNewDealStatusStream(s network.Stream) {
+	ctx := context.Background()
+	defer s.Close()
+
```
Comment (aarshkshah1992) [inline storagemarket/lp2pimpl/net.go:188]: Set a read deadline on the stream

File: storagemarket/lp2pimpl/net.go [hunk 4]
```diff
@@ -167,7 +177,120 @@ func (p *DealProvider) handleNewAskStream(s network.Stream) {
...
+		log.Errorw("failed to sign deal status response", "err", err)
+		return
+	}
+
+	response := &mktnet.DealStatusResponse{
+		DealState: *dealState,
+		Signature: *signature,
+	}
+
+	if err := cborutil.WriteCborRPC(s, response); err != nil {
```
Comment (aarshkshah1992) [inline storagemarket/lp2pimpl/net.go:222]: Set a write deadline

File: storagemarket/lp2pimpl/net.go [hunk 5]
```diff
@@ -167,7 +177,120 @@ func (p *DealProvider) handleNewAskStream(s network.Stream) {
...
+	}
+	signedPropCid := signedPropNd.Cid()
+
+	dealState := mktssm.ProviderDealState{
+		State:         toLegacyMarketsState(deal),
+		Proposal:      &deal.ClientDealProposal.Proposal,
+		ProposalCid:   &signedPropCid,
+		PublishCid:    deal.PublishCID,
+		DealID:        deal.ChainDealID,
+		FastRetrieval: true,
```
Comment (aarshkshah1992) [inline storagemarket/lp2pimpl/net.go:246]: why is this always true ?

File: storagemarket/lp2pimpl/net.go [hunk 6]
```diff
@@ -167,7 +177,120 @@ func (p *DealProvider) handleNewAskStream(s network.Stream) {
...
+	case dealcheckpoints.Transferred:
+		return mktssm.StorageDealCheckForAcceptance
+	case dealcheckpoints.Published:
+		return mktssm.StorageDealPublishing
+	case dealcheckpoints.PublishConfirmed:
+		return mktssm.StorageDealStaged
+	case dealcheckpoints.AddedPiece:
+		return mktssm.StorageDealAwaitingPreCommit
+	case dealcheckpoints.Complete:
+		return mktssm.StorageDealActive
```
Comment (aarshkshah1992) [inline storagemarket/lp2pimpl/net.go:292]: We currently mark deals as complete only when they fail

File: storagemarket/lp2pimpl/net.go [hunk 7]
```diff
@@ -167,7 +177,120 @@ func (p *DealProvider) handleNewAskStream(s network.Stream) {
...
+func toLegacyMarketsState(deal *types.ProviderDealState) mktssm.StorageDealStatus {
+	if deal.Err != "" {
+		return mktssm.StorageDealError
+	}
+
+	switch deal.Checkpoint {
+	case dealcheckpoints.Accepted:
+		return mktssm.StorageDealProposalAccepted
+	case dealcheckpoints.Transferred:
+		return mktssm.StorageDealCheckForAcceptance
```
Comment (aarshkshah1992) [inline storagemarket/lp2pimpl/net.go:284]: This should be `StorageDealVerifyData`

File: storagemarket/lp2pimpl/net.go [hunk 8]
```diff
@@ -167,7 +177,120 @@ func (p *DealProvider) handleNewAskStream(s network.Stream) {
...
+		Proposal:      &deal.ClientDealProposal.Proposal,
+		ProposalCid:   &signedPropCid,
+		PublishCid:    deal.PublishCID,
+		DealID:        deal.ChainDealID,
+		FastRetrieval: true,
+	}
+	return &dealState, nil
+}
+
+func (p *DealProvider) sign(ctx context.Context, dealState *mktssm.ProviderDealState) (*crypto.Signature, error) {
```
Comment (aarshkshah1992) [inline storagemarket/lp2pimpl/net.go:251]: are we doing this only for backward compat ?

File: storagemarket/lp2pimpl/net.go (hunk 6 above)
Comment (dirkmc) [reply storagemarket/lp2pimpl/net.go:308]: That case is taken care of on line 292:
```
	if deal.Err != "" {
		return mktssm.StorageDealError
	}
```

File: storagemarket/lp2pimpl/net.go (hunk 8 above)
Comment (dirkmc) [reply storagemarket/lp2pimpl/net.go:267]: Yep

File: db/deals.go (hunk 1 above)
Comment (dirkmc) [reply db/deals.go:77]: Added a test 👍

**Comment by dirkmc** [comment]
Closing as we will instead implement deal status v2

**Pull request closed without merging**